### PR TITLE
Documentation async

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Document Async. Only Documentation.
- Includes MultiSync, asyncCheck, waitFor, await, etc.
- Includes Examples.

:slightly_smiling_face:

### CODE REVIEW

**Comment by mratsim** [comment]
Maybe an async tutorial as well?

**Comment by juancarlospaco** [comment]
I agree with @mratsim that also a Tutorial can be nice, 
but Tutorials on asyncdispatch module feels weird and probably hard to discover for new users.

Maybe Tutorials improvement can be part of another PR, to keep diff easy to review and merge.

**Comment by Araq** [comment]
IMHO this should be an async tutorial and the asyncdispatch modules should link to the tutorial.

**Comment by juancarlospaco** [comment]
We all agree on that.
But if you want to close this one close it, one way or another users will not complain for too much Documentation. :P

File: lib/pure/asyncdispatch.nim [hunk 1]
```diff
@@ -136,10 +156,51 @@
...
+##
+## For more examples take a look at the documentation for the modules implementing
 ## asynchronous IO. A good place to start is the
 ## `asyncnet module <asyncnet.html>`_.
 ##
+##
+## waitFor
+## -------
+##
+## ``waitFor`` is similiar to ``asyncCheck`` but it will wait for the ``Future``.
```
[the remark this answers is not in the document]
Comment (dom96) [reply lib/pure/asyncdispatch.nim]: This is a bad description though. ``waitFor`` blocks the current thread and runs the async IO loop until the future finishes, it's not really comparable to ``asyncCheck``.

File: lib/pure/asyncdispatch.nim [hunk 2]
```diff
@@ -136,10 +156,51 @@
...
+##
+##   waitFor example()
+##
+## await
+## -----
+##
+## ``await`` will wait for the ``Future``, and gives you the contents inside the ``Future``,
+## ``await`` only available inside a function with the ``{.async.}`` or ``{.multisync.}`` Pragmas.
+##
+## The usage and features are the same of ``waitFor`` (mentioned above).
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: usage is the same, I don't understand what features there are to speak of which are the same.

File: lib/pure/asyncdispatch.nim [hunk 3]
```diff
@@ -136,10 +156,51 @@
...
+##
+## Comparison
+## ----------
+##
+## ============================ ============ ========= =======
+##  Description                  asyncCheck   waitFor   await
+## ============================ ============ ========= =======
+## Waits the Future              No           Yes       Yes
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]:
```suggestion
## Waits for the Future to complete              No           Yes       Yes
```

File: lib/pure/asyncdispatch.nim [hunk 4]
```diff
@@ -160,6 +221,41 @@
 ## graph.
 ##
 ##
+## Multisync
+## =========
+##
+## The ``{.multisync.}`` Pragma is similar to the ``{.async.}`` Pragma,
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: Pragma shouldn't be capitalised...

File: lib/pure/asyncdispatch.nim [hunk 5]
```diff
@@ -160,6 +221,41 @@
...
+## ``{.multisync.}`` Pragma converts a procedure to an asynchronous *and* synchronous procedure,
+## then the same function can be used on asynchronous *and* synchronous code,
+## you can think of it as a function that is async or sync depending on how you use it.
+##
+## The usage and features are the same of ``{.async.}`` (mentioned above). Example:
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: again with the "features", that tells the reader nothing and is misleading.

File: lib/pure/asyncdispatch.nim [hunk 6]
```diff
@@ -160,6 +221,41 @@
...
+##
+##   type
+##     Client = HttpClient
+##     AsyncClient = AsyncHttpClient
+##
+##   proc multisync_function(this: Client | AsyncClient): Future[string] {.multisync.} =
+##     const url = "https://nim-lang.org"
+##     result =
+##       when this is AsyncClient: await newAsyncHttpClient().getContent(url)
+##       else: newHttpClient().getContent(url)
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: This is a bad example, create the http client and assign it to a variable instead

File: lib/pure/asyncdispatch.nim [hunk 7]
```diff
@@ -160,6 +221,41 @@
...
+##
+##   proc async_example() {.async.} =
+##     echo await AsyncClient().multisync_function()
+##
+##   waitFor async_example()            # Async
+##
+##   echo Client().multisync_function() # Sync
+##
+## This is one of the reasons Nim wont use syntax like ``async proc fetch() =`` like other languages,
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: *won't

File: lib/pure/asyncdispatch.nim [hunk 8]
```diff
@@ -160,6 +221,41 @@
...
+## also you dont have to repeat your procedure twice like ``async proc aio_fetch() =`` and ``proc fetch() =``,
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: *don't...

File: lib/pure/asyncdispatch.nim [hunk 9]
```diff
@@ -160,6 +221,41 @@
...
+## The usage and features are the same of ``{.async.}`` (mentioned above). Example:
+##
+## .. code-block:: nim
+##   import asyncdispatch, httpclient
+##
+##   type
+##     Client = HttpClient
+##     AsyncClient = AsyncHttpClient
+##
+##   proc multisync_function(this: Client | AsyncClient): Future[string] {.multisync.} =
```
Comment (dom96) [inline lib/pure/asyncdispatch.nim]: camelCase is the convention

File: lib/pure/asyncdispatch.nim (hunk 7 above)
Comment (dom96) [reply lib/pure/asyncdispatch.nim]: How is *this* one of the reasons?

File: lib/pure/asyncdispatch.nim [hunk 10]
```diff
@@ -133,13 +153,43 @@
...
+##
+## Comparison
+## ----------
+##
+## ================================ ============ ========= =======
+##  Description                      asyncCheck   waitFor   await
+## ================================ ============ ========= =======
+## Waits for the Future to complete  No           Yes       Yes
+## Ignores the Future                Yes          No        No
+## Returns result inside Future      No           No        Yes
```
Comment (treeform) [inline lib/pure/asyncdispatch.nim:188]: "Returns result inside Future"  waitFor should be yes?

Example:

```nim
import asyncdispatch

proc foo(): Future[int] {.async.} =
  return 42

echo waitFor foo()
```

**Comment by treeform** [comment]
I really like this. I wish this was there when I started to learn about async. The async only clicked for me when I learned that:
* asyncCheck - is like `spawn` in other languages - it starts a new async-thread and branches execution. 
* waitFor and await - are basically same thing, one you use inside async code the other one you use outside async code. If compiler complains just switch one to the other :)

Off-topic, but if it was up to me I would:
* Rename asyncCheck -> spawn 
* Make await act like waitFor when outside {.async.} functions
* Wrap result in a Future automatically when using {.async.}

**Comment by juancarlospaco** [comment]
...but it does wrap the result on a `Future`, even if you do not declare an explicit return it returns `Future[void]`. :thinking:

**Comment by disruptek** [comment]
`return` wraps its argument but `result` is always `Future`.

**Comment by treeform** [comment]
Sorry I meant do that in the definitions of the proc. I know it does that
on the actual return, I wish it would do that in the signature as well to
match.

**Comment by disruptek** [comment]
The problem with that, assuming you didn't have the burden of code in the wild, is that it's hiding even more magic.  I wish `return` behaved as expected, instead.

**Comment by treeform** [comment]
It is too late to change how {.async.} works now. Sorry for derailing the discussion. I think the docs are good.

**Comment by narimiran** [comment]
> This doesn't belong in the manual as it's not a language feature, it should probably be in a separate document 

> Maybe an async tutorial as well?

> IMHO this should be an async tutorial and the asyncdispatch modules should link to the tutorial.

> We all agree on that.

So, if we all agree on that, can we close this PR and use its content as a basis for `doc/tut4.rst`, which will be async tutorial?

**Comment by juancarlospaco** [comment]
Feel free to close if is the way forward.

**Comment by narimiran** [comment]
> Feel free to close if is the way forward.

I think it is. Now waiting for your async tutorial :)

**Comment by ringabout** [comment]
What's the progress of this tutorial?

**Comment by Araq** [comment]
We don't have a tutorial so we should take the doc improvements. Please continue your work @juancarlospaco !

File: lib/pure/asyncdispatch.nim (hunk 10 above)
Comment (xflywind) [reply lib/pure/asyncdispatch.nim:188]: This also fixes https://github.com/nim-lang/Nim/issues/5021

**Comment by ringabout** [comment]
This also fixes https://github.com/nim-lang/Nim/issues/5021

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/19738

**Pull request closed without merging**